autoware_lanelet2_map_validator/src/validators/intersection/right_of_way_for_virtual_traffic_lights.cpp

```diff
@@ -0,0 +1,163 @@
+
+lanelet::validation::Issues
+RightOfWayForVirtualTrafficLightsValidator::check_right_of_way_for_virtual_traffic_lights(
+  const lanelet::LaneletMap & map)
+{
+  lanelet::validation::Issues issues;
+
+  auto traffic_rules = lanelet::traffic_rules::TrafficRulesFactory::create(
+    lanelet::Locations::Germany, lanelet::Participants::Vehicle);
+  auto routing_graph = lanelet::routing::RoutingGraph::build(map, *traffic_rules);
+
+  for (const auto & lanelet : map.laneletLayer) {
+    const auto virtual_traffic_light_elems =
+      lanelet.regulatoryElementsAs<lanelet::autoware::VirtualTrafficLight>();
+
+    if (virtual_traffic_light_elems.empty()) {
+      continue;
+    }
+
+    const auto right_of_way_elems = lanelet.regulatoryElementsAs<lanelet::RightOfWay>();
+
+    if (right_of_way_elems.empty()) {
+      // Issue-001: Lanelet with virtual_traffic_light missing right_of_way reference
+      issues.emplace_back(construct_issue_from_code(issue_code(this->name(), 1), lanelet.id()));
+      continue;
+    }
+
+    if (right_of_way_elems.size() > 1) {
+      // Issue-002: Multiple right_of_way regulatory elements in the same lanelet
+      issues.emplace_back(construct_issue_from_code(issue_code(this->name(), 2), lanelet.id()));
+      continue;
+    }
+
+    const auto right_of_way_elem = right_of_way_elems.front();
+
+    bool is_set_as_right_of_way = false;
+    const auto right_of_way_lanelets =
+      right_of_way_elem->getParameters<lanelet::ConstLanelet>(lanelet::RoleName::RightOfWay);
+
+    if (right_of_way_lanelets.size() != 1) {
+      // Issue-003: right_of_way regulatory element should have exactly one right_of_way role
+      issues.emplace_back(
+        construct_issue_from_code(issue_code(this->name(), 3), right_of_way_elem->id()));
+      continue;
+    }
+
+    for (const auto & row_lanelet : right_of_way_lanelets) {
+      if (row_lanelet.id() == lanelet.id()) {
+        is_set_as_right_of_way = true;
+        break;
+      }
+    }
+
+    if (!is_set_as_right_of_way) {
+      // Issue-004: right_of_way regulatory element doesn't set this lanelet as right_of_way role
+      issues.emplace_back(
+        construct_issue_from_code(issue_code(this->name(), 4), right_of_way_elem->id()));
+      continue;
+    }
+
+    const auto yield_lanelets =
+      right_of_way_elem->getParameters<lanelet::ConstLanelet>(lanelet::RoleName::Yield);
+
+    std::vector<lanelet::ConstLanelet> conflicting_lanelets;
+    for (const auto & other_lanelet : map.laneletLayer) {
+      if (other_lanelet.id() == lanelet.id()) {
+        continue;
+      }
+
+      const auto relation = routing_graph->routingRelation(lanelet, other_lanelet);
+      if (relation == lanelet::routing::RelationType::Conflicting) {
+        conflicting_lanelets.push_back(other_lanelet);
+      }
+    }
+
+    std::set<lanelet::Id> conflicting_ids;
+    for (const auto & conflicting_lanelet : conflicting_lanelets) {
+      conflicting_ids.insert(conflicting_lanelet.id());
+    }
```

Comment:
I think using `routing_graph->conflicting(lanelet)` is better so that you don't have to iterate through the lanelet later.
It's a bit tricky because it returns a `lanelet::ConstLaneletOrArea` object but its not that hard.

```suggestion
    const auto conflicting_primitives = routing_graph->conflicting(lanelet);

    std::set<lanelet::Id> conflicting_ids;
    for (const auto & primitive : conflicting_primitives) {
      if(primitive.isArea()) {
        continue;
      }
      conflicting_ids.insert(primitive.id());
    }